yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -62,10 +62,19 @@ private[spark] class ApplicationMaster(
     .asInstanceOf[YarnConfiguration]
   private val isClusterMode = args.userClass != null
 
-  // Default to numExecutors * 2, with minimum of 3
-  private val maxNumExecutorFailures = sparkConf.getInt("spark.yarn.max.executor.failures",
-    sparkConf.getInt("spark.yarn.max.worker.failures",
-      math.max(sparkConf.getInt("spark.executor.instances", 0) *  2, 3)))
+  // Default to twice the number of executors (twice the maximum number of executors if dynamic
+  // allocation is enabled), with a minimum of 3.
+  val defaultKey =
+    if (Utils.isDynamicAllocationEnabled(sparkConf)) {
+      "spark.dynamicAllocation.maxExecutors"
+    } else {
+      "spark.executor.instances"
+    }
+  val effectiveNumExecutors = sparkConf.getInt(defaultKey, 0)
+  val defaultMaxNumExecutorFailures = math.max(3, 2 * effectiveNumExecutors)
```

Comment:
you addressed my comment quickly, but these should not be public or outside the scope of `maxNumExecutorFailures`. What I meant is more like:
```
// Default to twice ...
private val maxNumExecutorFailures: Int = {
  val defaultKey = ...
  sparkConf.getInt(...)
}
```